ai.py

```diff
@@ -1,98 +1,145 @@
+                moves.append(Move(src, dst))
+        elif piece == 2:
+            for dir in knight_directions:
+                dst = src + dir
 
-                dst = Position(x - 1, y + player_sign)
-                if board.can_capture_space(dst, player_sign):
+                if board.can_move_space(dst):
+                    moves.append(Move(src, dst))
+                elif board.can_capture_space(dst, player_sign):
                     moves.append(Move(src, dst))
-            elif piece == 2:
-                for dir in [
-                    Position(1, 2), Position(-1, 2),
-                    Position(2, -1), Position(2, 1),
-                    Position(1, -2), Position(-1, -2),
-                    Position(-2, -1), Position(-2, 1)
-                ]:
-                    dst = src + dir
+        elif piece == 3 or piece == 5:
+            for dir in diagonal_directions:
+                for dist in range(1, 8):
+                    dst = src + dist * dir
 
                     if board.can_move_space(dst):
                         moves.append(Move(src, dst))
                     elif board.can_capture_space(dst, player_sign):
                         moves.append(Move(src, dst))
-            elif piece == 3 or piece == 5:
-                for dir in [
-                    Position(1, 1), Position(-1, 1),
-                    Position(1, -1), Position(-1, -1)
-                ]:
-                    for dist in range(1, 8):
-                        dst = src + dist * dir
-
-                        if board.can_move_space(dst):
-                            moves.append(Move(src, dst))
-                        elif board.can_capture_space(dst, player_sign):
-                            moves.append(Move(src, dst))
-                            break
-                        else:
-                            break
-            elif piece == 4 or piece == 5:
-                for dir in [
-                    Position(1, 0), Position(-1, 0),
-                    Position(0, 1), Position(0, -1)
-                ]:
-                    for dist in range(1, 8):
-                        dst = src + dist * dir
-
-                        if board.can_move_space(dst):
-                            moves.append(Move(src, dst))
-                        elif board.can_capture_space(dst, player_sign):
-                            moves.append(Move(src, dst))
-                            break
-                        else:
-                            break
-            elif piece == 6:
-                for dir in [
-                    Position(1, 0), Position(-1, 0),
-                    Position(0, 1), Position(0, -1),
-                    Position(1, 1), Position(-1, 1),
-                    Position(1, -1), Position(-1, -1)
-                ]:
-                    dst = src + dir
+                        break
+                    else:
+                        break
+        
+        if piece == 4 or piece == 5:
```

Comment:
This needs to be if instead of elif so that queen will see this code as well as the code that moves queen diagonally.